packages/financial-templates-lib/src/clients/InsuredBridgeL2Client.ts

```diff
@@ -122,14 +128,113 @@ export class InsuredBridgeL2Client {
+      this.bridgeDepositBox.options.jsonInterface,
+      this.bridgeDepositAddress,
+      "WhitelistToken",
+      blockSearchConfig
+    );
+    // TODO: Should we throw, or even send a warning?
+    if (eventsData.missingEvents.length > 0) {
+      // TODO: Do something with eventsData.missingEvents array?
+      throw new Error(
+        `${eventsData.missingEvents.length} WhitelistToken events found in fallback l2 provider not found in all l2 web3 providers`
+      );
+    }
+
+    // All events were found in all providers, can return any of the event data arrays
+    return eventsData.events;
   }
 
+  /**
+   * Fetches specified contract event data for all input web3 providers. Returns false if any of the events found with
+   * one provider are NOT matched exactly in ALL of the other providers' event arrays.
+   * @param web3s Web3 providers to check for target event.
+   * @param contractAbi Contract ABI to query target event on.
+   * @param contractAddress Contract address to query target event on.
+   * @param eventName Name of target event.
+   * @param blockSearchConfig Target event search config
+   * @returns Object containing success of event query, missing events not found in all providers, and all event data
+   */
+  async getEventsForMultipleProviders(
+    web3s: Web3[],
+    contractAbi: any[],
+    contractAddress: string,
+    eventName: string,
+    blockSearchConfig: BlockSearchConfig
+  ): Promise<{ missingEvents: EventData[]; events: EventData[] }> {
+    const allProviderEvents = await Promise.all(
+      web3s.map((_web3) => {
+        const _contract = new _web3.eth.Contract(contractAbi, contractAddress);
+        return _contract.getPastEvents(eventName, blockSearchConfig);
+      })
+    );
+
+    // Associate each event with a key uniquely identifying the event. We'll use this key in the next step to determine
+    // which events were returned by all providers.
+    type UniqueEventData = { [uniqueEventKey: string]: EventData };
+    const uniqueEventsForProvider: UniqueEventData[] = [];
+    // [index of web3Provider => {eventKey => event}]
+
+    const _getUniqueEventKey = (event: EventData): string => {
+      return JSON.stringify({
+        transactionHash: event.transactionHash,
+        transactionIndex: event.transactionIndex,
+        logIndex: event.logIndex,
+        returnValues: event.returnValues,
+        address: event.address,
+      });
+    };
+    allProviderEvents.forEach((eventDataForProvider, i) => {
+      uniqueEventsForProvider[i] = {};
+      eventDataForProvider.forEach((event) => {
+        const uniqueEventKey = _getUniqueEventKey(event);
+        uniqueEventsForProvider[i][uniqueEventKey] = event;
+      });
+    });
+
+    // Store only the events returned by ALL providers.
+    const eventKeysReturnedByAllProviders: string[] = [];
+    const missingEventKeys: string[] = [];
+    Object.keys(uniqueEventsForProvider[0]).forEach((eventKeysForFirstProvider: string) => {
```

Comment:
What happens if an event is missing in the 0th provider?

If you agree that this case isn't covered, I might have a slight tweak that would make it work. You could add a union map that you build in the first loop. If you encounter any events that aren't in that map, you add it in the first loop.

Then in this loop, you can loop over all the elements in that union map so you know that you're covering everything.

Reply:
Ah yes I agree with this optimization. It adds an additional struct but no more loops